[libs] Bug Fix on Bitmap Allocation

ppenna opened the pull request:
## Description

This PR enhances error handling and logging across several memory management modules, optimizes bitmap allocation logic, and improves code clarity. The most significant changes include adding detailed error logging, refactoring the bitmap allocation methods for efficiency, and introducing helper functions for internal bitmap operations.

### Enhanced error handling and logging:
* Added detailed error logging for allocation, freeing, and booking operations in `FrameAllocator`, `KpoolInner`, and `Vmem` to provide better debugging information.

### Bitmap allocation improvements:
* Refactored `Bitmap::alloc` to delegate to `Bitmap::alloc_range` for code reuse and consistency.
* Optimized `Bitmap::alloc_range` to traverse the bitmap more efficiently by introducing fast paths for aligned and multi-byte allocations. Added checks to ensure allocations do not exceed capacity.

### Code clarity and helper functions:
* Renamed `len` parameter to `number_of_bits` in `Bitmap::new` for better readability and added validation to ensure lengths are multiples of 8.
* Introduced `Bitmap::index_unchecked` to simplify internal operations by calculating word and bit positions without bounds checking.

### Miscellaneous:
* Updated `UserStackAllocator` to simplify bitmap initialization.
* Enabled the `random` feature in test configurations for bitmap testing.

ppenna committed:
[kernel] E: Improve Error Logging in Memory Module

In this commit I improved error logging in the memory management module
to print an error message as soon as an errors is detected.

ppenna committed:
[libs] E: Bitmap Creation Based on Number of Bits

ppenna committed:
[libs] E: Make alloc alias for alloc_range

ppenna committed:
[tests] E: Additional Test Cases for Bitmap

This commit introduces additional test cases for the bitmap structure:

- test_alloc_range_too_large()
- test_alloc_range_zero()
- test_alloc_range_across_word_boundary()
- test_alloc_random_ranges()
- test_alloc_random_ranges_in_partial_bitmap()

The following bugs are uncovered:

- test_alloc_range_across_word_boundary() fails
- test_alloc_random_ranges() fails
- test_alloc_random_ranges_in_partial_bitmap() fails

ppenna committed:
[libs] B: Buggy Bitmap Allocation

This commit fixes buggy bitmap allocation.

ppenna committed:
[libs] E: Faster Bitmap Allocation (#574)

This commit resolves https://github.com/nanvix/nanvix/issues/574.

ppenna commented:
Issue https://github.com/nanvix/nanvix/issues/576 was triggered when running the CI.

esaurez commented on src/libs/bitmap/src/lib.rs:
```diff
@@ -170,40 +157,61 @@ impl Bitmap {
+        // Traverse the bitmap until the last possible starting bit.
+        while start <= self.number_of_bits - size {
+            // Check for fast skip/ path.
+            let word: usize = start / u8::BITS as usize;
```
Maybe move inside the if is_aligned. The variable is not used anywhere else.

ppenna replied:
Done.

esaurez commented on src/libs/bitmap/src/lib.rs:
```diff
@@ -170,40 +157,61 @@ impl Bitmap {
+                // Fast skip: if the starting word is full, skip to the next word.
+                if self.bits[word] == u8::MAX {
+                    // Jump to next byte boundary.
+                    let next_start: usize = (word + 1) * u8::BITS as usize;
```
NIT, wouldn't it be simpler to just do start + u8::BITS instead of the multiplication, it makes it more explicit that you are moving one word ahead.

ppenna replied:
Makes sense. Done.

ppenna committed:
[kernel] E: Improve Tracing on Thread Creation

This commit improves tracing on thread creation to help us to find the
root cause of https://github.com/nanvix/nanvix/issues/576.

ppenna commented on src/libs/bitmap/src/lib.rs:
```diff
@@ -170,40 +157,61 @@ impl Bitmap {
-            let reason: &str = "size out of bounds";
-            return Err(Error::new(ErrorCode::InvalidArgument, reason));
+        // Check if allocation exceeds the bitmap capacity.
+        if self.usage > self.number_of_bits - size {
```
This would cause an overflow.

ppenna committed:
[ci] E: Increase Workflow Timeout for Hyperlight
